Sources/GRPCReflectionService/Server/ReflectionService.swift

```diff
@@ -0,0 +1,184 @@
+/*
+ * Copyright 2023, gRPC Authors All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import DequeModule
+import Foundation
+import GRPC
+import SwiftProtobuf
+
+@available(macOS 10.15, iOS 13, tvOS 13, watchOS 6, *)
+public class ReflectionService: CallHandlerProvider {
+  public var serviceName: Substring
+  private let reflectionService: ReflectionServiceProvider
+
+  public init(fileDescriptors: [Google_Protobuf_FileDescriptorProto]) throws {
+    self.reflectionService = try ReflectionServiceProvider(fileDescriptorProtos: fileDescriptors)
+    self.serviceName = self.reflectionService.serviceName
+  }
+
+  public func handle(
+    method name: Substring,
+    context: GRPC.CallHandlerContext
+  ) -> GRPC.GRPCServerHandlerProtocol? {
+    self.reflectionService.handle(method: name, context: context)
+  }
+}
+
+internal struct ReflectionServiceData: Sendable {
+  internal struct FileDescriptorProtoData: Sendable {
+    internal var serializedFileDescriptorProto: Data
+    internal var dependencyFileNames: [String]
+  }
+
+  internal var fileDescriptorDataByFilename: [String: FileDescriptorProtoData]
+  internal var serviceNames: [String]
+
+  internal init(fileDescriptors: [Google_Protobuf_FileDescriptorProto]) throws {
+    self.serviceNames = []
+    self.fileDescriptorDataByFilename = [:]
+    do {
+      for fileDescriptorProto in fileDescriptors {
+        let protoDataObj = FileDescriptorProtoData(
+          serializedFileDescriptorProto: try fileDescriptorProto.serializedData(),
+          dependencyFileNames: fileDescriptorProto.dependency
+        )
+        self.fileDescriptorDataByFilename[fileDescriptorProto.name] = protoDataObj
+        self.serviceNames.append(contentsOf: fileDescriptorProto.service.map { $0.name })
+      }
+    } catch {
+      throw GRPCStatus(
+        code: .invalidArgument,
+        message:
+          "The \(fileDescriptors[self.fileDescriptorDataByFilename.count].name) could not be serialized."
+      )
+    }
+  }
+
+  public func getSerializedFileDescriptorProtos(fileName: String) throws -> [Data] {
```

Comment:
This should be `internal`

Reply:
I also wonder if this is a useful name: it's not clear what protos it returns.

Something like `serialisedFileDescriptorProtosForDependenciesOfFile(named fileName: String)` is more verbose but at least tells us we're returning the dependencies of the file